etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -173,28 +222,181 @@ process_parcel_file <- function(s3_bucket_uri,
+    # rectangle
+    spatial_df_rec <- spatial_df_merged %>%
+      st_set_geometry("geometry") %>%
+      st_drop_geometry() %>%
+      st_set_geometry("geometry_3435") %>%
+      mutate(parcel_area = st_area(geometry_3435)) %>%
+      st_minimum_rotated_rectangle() %>%
+      mutate(
+        shp_parcel_mrr_area_ratio = units::drop_units(
+          st_area(geometry_3435) / parcel_area
+        ),
+        # Sometimes the ratio here is very slightly less than 1, even though
+        # that should be impossible. This seems to happen mostly for extremely
+        # long, skinny triangle parcels. The error is small enough that it's
+        # probably fine to ignore for now
+        shp_parcel_mrr_area_ratio = ifelse(
+          shp_parcel_mrr_area_ratio < 1,
+          1,
+          shp_parcel_mrr_area_ratio
+        )
+      )
+
+    # Calculate the ratio of the sides of the minimum bounding rectangle
+    spatial_mat_rec_coords <- spatial_df_rec %>%
+      st_coordinates() %>%
+      as.data.table()
 
-    # Sort by year, town code, and PIN10 for better compression
-    spatial_df_merged <- spatial_df_merged %>%
+    spatial_mat_rec_calc <- spatial_mat_rec_coords[
+      ,
+      edge_len := sqrt(
+        (X - data.table::shift(X, type = "lag")) ^ 2 +
+          (Y - data.table::shift(Y, type = "lag")) ^ 2
+      ),
+      by = c("L1", "L2")
+    ][
+      ,
+      .(
+        shp_parcel_mrr_side_ratio = max(edge_len, na.rm = TRUE) /
+          min(edge_len, na.rm = TRUE)
+      ),
+      by = "L2"
+    ]
+    tictoc::toc()
+
+    tictoc::tic(paste("Finalized and wrote to file:", file_year))
+    # Merge all calculated features back into the main dataframe
+    # then sort by year, town code, and PIN10 for better compression
+    spatial_df_final <- spatial_df_merged %>%
+      cbind(
+        spatial_df_rec %>%
+          select(shp_parcel_mrr_area_ratio) %>%
+          st_drop_geometry(),
+        spatial_mat_rec_calc[, 2],
+        spatial_mat_calc_vert[, 2:5]
+      ) %>%
       ungroup() %>%
-      arrange(year, town_code, pin10)
+      arrange(year, town_code, pin10) %>%
+      relocate(starts_with("shp_"), .after = "y_3435") %>%
+      relocate(
+        c(shp_parcel_centroid_dist_ft_sd, shp_parcel_interior_angle_sd),
+        .after = "shp_parcel_edge_len_ft_sd"
+      )
+
+    # Check that the final number of distinct, well-formed parcels is close to
+    # the same as the number in the raw parcel file
+    num_parcel_raw <- spatial_df_raw %>%
+      rename_with(tolower) %>%
+      filter(!is.na(pin10), nchar(pin10) == "10", !st_is_empty(geometry)) %>%
+      st_drop_geometry() %>%
+      summarize(n = n_distinct(pin10)) %>%
+      pull(n)
+    num_parcel_final <- spatial_df_final %>%
+      st_drop_geometry() %>%
+      summarize(n = n_distinct(pin10)) %>%
+      pull(n)
+    if (abs(num_parcel_raw - num_parcel_final) > 10) {
+      stop("More than 10 PINs dropped during geometry processing!")
+    }
```

Comment:
**[Thought, non-blocking]** Good thinking adding this test! It makes me wonder more broadly about how we should define QC tests on non-dbt transformation scripts like this one. Maybe running tests like this at the end of the script is the best solution? Or maybe we split tests out into a function defined in a different module, and then import and call that function at the end of this script for better separation of concerns? Either way, I support the current solution in the immediate term; this is just a thought to keep in mind if we find ourselves adding more QC tests like this one to non-dbt transformations.

Reply:
Let's leave it for now, since this script run rarely enough that I don't think investing lots of time into them is a good idea. However, this is definitely something to think about when we look at #518.